### PR TITLE
Webrtc use RTCIceCandidateInit messages with backend

### PR DESCRIPTION
**Pull request by sdb9696**
## Proposed change

Pass the `RTCIceCandidateInit` message between the frontend and backend to allow setting the sdpMid and sdpMLineIndex values.

[core PR](https://github.com/home-assistant/core/pull/129879)

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by sdb9696** [comment]
> Can you link the associated core PR?

That's done

File: src/components/ha-web-rtc-player.ts [hunk 1]
```diff
@@ -248,11 +248,13 @@ class HaWebRtcPlayer extends LitElement {
...
       this._sessionId = event.session_id;
       this._candidatesList.forEach((candidate) =>
+        // sdpMLineIndex is always populated
         addWebRtcCandidate(
           this.hass,
           this.entityid!,
           event.session_id,
-          candidate
+          candidate.candidate,
+          candidate?.sdpMLineIndex || 0
```
Comment (bramkragten) [inline src/components/ha-web-rtc-player.ts:257]:
```suggestion
          candidate.sdpMLineIndex || 0
```

**Review by bramkragten — DISMISSED** [review]

**Comment by sdb9696** [comment]
Hi @bramkragten, since you approved this I have updated the messaging to pass the `RTCIceCandidate` message between the front end the core rather than just adding `sdp_m_line_index`. This makes the solution more in line with the standard which could accept `sdpMid` as well.

**Comment by bramkragten** [comment]
@sdb9696 I like that, but the types that are used dont match, the type in frontend had a lot more than the model in the backend.

Create a separate type in the frontend that matches the type of the backend.

<https://developer.mozilla.org/en-US/docs/Web/API/RTCIceCandidate>

<https://github.com/home-assistant-libs/python-webrtc-models/pull/2/files#diff-2cb5148764f0e6bbc48bc92652db6485ce1dd68cfa48e37dd19ab9716c4c600eR58-R64>

**Comment by sdb9696** [comment]
> @sdb9696 I like that, but the types that are used dont match, the type in frontend has a lot more than the model in the backend.
> 
> Create a separate type in the frontend that matches the type of the backend.
> 
> [developer.mozilla.org/en-US/docs/Web/API/RTCIceCandidate](https://developer.mozilla.org/en-US/docs/Web/API/RTCIceCandidate)
> 
> [home-assistant-libs/python-webrtc-models#2 (files)](https://github.com/home-assistant-libs/python-webrtc-models/pull/2/files#diff-2cb5148764f0e6bbc48bc92652db6485ce1dd68cfa48e37dd19ab9716c4c600eR58-R64)

Hi @bramkragten, using the same type seems to be consistent with how the [RTCConfiguration](https://github.com/home-assistant/frontend/blob/dev/src/components/ha-web-rtc-player.ts#L116:L125) is being used, i.e. the `webrtc-models` implementation is not defining all the fields in the spec but the message from the backend is being used to construct the `RTCConfiguration` directly. At the moment the web socket just ignores the extra fields in the message it gets back from the frontend. Do you think that's an issue? Another approach could be to implement all the ice candidate spec in `webrtc-models` but it seems unnecessary as all the other fields are derived from the `candidate` string so we'll never need them.

**Comment by bramkragten** [comment]
The difference there is that for `RTCConfiguration` every property is optional, so it is valid.

```ts
interface RTCConfiguration {
    bundlePolicy?: RTCBundlePolicy;
    certificates?: RTCCertificate[];
    iceCandidatePoolSize?: number;
    iceServers?: RTCIceServer[];
    iceTransportPolicy?: RTCIceTransportPolicy;
    rtcpMuxPolicy?: RTCRtcpMuxPolicy;
}
```

For `RTCIceCandidate` there is no optional property, they can be null, but not omitted. Also there should be a `toJSON` function.

```ts
interface RTCIceCandidate {
    /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/RTCIceCandidate/address) */
    readonly address: string | null;
    /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/RTCIceCandidate/candidate) */
    readonly candidate: string;
    /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/RTCIceCandidate/component) */
    readonly component: RTCIceComponent | null;
    /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/RTCIceCandidate/foundation) */
    readonly foundation: string | null;
    /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/RTCIceCandidate/port) */
    readonly port: number | null;
    /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/RTCIceCandidate/priority) */
    readonly priority: number | null;
    /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/RTCIceCandidate/protocol) */
    readonly protocol: RTCIceProtocol | null;
    /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/RTCIceCandidate/relatedAddress) */
    readonly relatedAddress: string | null;
    /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/RTCIceCandidate/relatedPort) */
    readonly relatedPort: number | null;
    /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/RTCIceCandidate/sdpMLineIndex) */
    readonly sdpMLineIndex: number | null;
    /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/RTCIceCandidate/sdpMid) */
    readonly sdpMid: string | null;
    /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/RTCIceCandidate/tcpType) */
    readonly tcpType: RTCIceTcpCandidateType | null;
    /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/RTCIceCandidate/type) */
    readonly type: RTCIceCandidateType | null;
    /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/RTCIceCandidate/usernameFragment) */
    readonly usernameFragment: string | null;
    /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/RTCIceCandidate/toJSON) */
    toJSON(): RTCIceCandidateInit;
}
```

**Comment by sdb9696** [comment]
Ah, good point. So I think I'll need to extend the RTCIceCandidate interface. Will check with @edenhaus.

**Comment by sdb9696** [comment]
Ok have now more fully implemented `RTCIceCandidate` in the [webrtc-models PR](https://github.com/home-assistant-libs/python-webrtc-models/pull/2).

`to_json` will return the `RTCIceCandidateInit` as per the spec.

**Comment by sdb9696** [comment]
This is now updated to use the `RTCIceCandidateInit` for messaging and supports just the candidate being received to avoid breaking change.

**Review by bramkragten — APPROVED** [review]

**Comment by edenhaus** [comment]
Core PR is merged, so we can merge this one

**Pull request merged**